psi4/src/psi4/libsapt_solver/fdds_disp.cc

```diff
@@ -481,5 +542,547 @@ SharedMatrix FDDS_Dispersion::form_unc_amplitude(std::string monomer, double ome
+    auto YarQ = std::make_shared<Matrix>("YarQ", maxo * nvir, naux);
+    auto QYarQ = std::make_shared<Matrix>("QYarQ", maxo * nvir, naux);
+    auto YarQL = std::make_shared<Matrix>("YarQL", maxo * nvir, naux);
+
+    // => Target <= //    
+    std::vector<SharedMatrix> ret;
+    for (size_t i = 0; i < 5; i++) {
+        ret.push_back(std::make_shared<Matrix>(naux, naux));
+        ret[i]->zero();
+    }
+
+    // => Pointers <= //
+
+    std::vector<double**> retp;
+    for (size_t i = 0; i < 5; i++) {
+        retp.push_back(ret[i]->pointer());
+    }    
+
+    double** arQp = arQ->pointer();
+    double** arQLDp = arQLD->pointer();
+    double** XarQp = XarQ->pointer();
+    double** YarQp = YarQ->pointer();
+    double** YarQLp = YarQL->pointer();
+    double** QXarQp = QXarQ->pointer();
+    double** QYarQp = QYarQ->pointer();
+
+    // => Master Loop <= //
+
+    for (size_t astart = 0; astart < nocc; astart += maxo) {
+        size_t nablock = (astart + maxo >= nocc ? nocc - astart : maxo);
+
+        size_t navir = nablock * nvir;
+
+        dfh_->fill_tensor(arQ_name, arQ, {astart, astart + nablock});
+        dfh_->fill_tensor(XarQ_name, XarQ, {astart, astart + nablock});
+        dfh_->fill_tensor(QXarQ_name, QXarQ, {astart, astart + nablock});
+        dfh_->fill_tensor(YarQ_name, YarQ, {astart, astart + nablock});
+        dfh_->fill_tensor(QYarQ_name, QYarQ, {astart, astart + nablock});
+
+        ret.push_back(arQ);
+        ret.push_back(XarQ);
+        ret.push_back(QXarQ);
+        ret.push_back(YarQ);
+        ret.push_back(QYarQ);
+
+        
+        // X <- X + Y, Y <- Y - X
+        XarQ->axpy(1.0, YarQ);
+        YarQ->scale(2.0);
+        YarQ->axpy(-1.0, XarQ);
+
+        QXarQ->axpy(1.0, QYarQ);
+        QYarQ->scale(2.0);
+        QYarQ->axpy(-1.0, QXarQ);
+
+        arQLD->copy(arQ);
+        YarQL->copy(YarQ);
+
+#pragma omp parallel for collapse(2)
+        for (size_t a = 0; a < nablock; a++) {
+            for (size_t r = 0; r < nvir; r++) {
+                double ll = Larp[astart + a][r];
+                double ld = LDarp[astart + a][r];
+#pragma omp simd
+                for (size_t Q = 0; Q < naux; Q++) {
+                    arQLDp[a * nvir + r][Q] *= ld;
+                    YarQLp[a * nvir + r][Q] *= ll;
+                }
+            }
+        }
+
+        C_DGEMM('T', 'N', naux, naux, navir, 1.0, arQLDp[0], naux, arQp[0], naux, 1.0, retp[0][0], naux);
+        C_DGEMM('T', 'N', naux, naux, navir, 1.0, arQLDp[0], naux, QXarQp[0], naux, 1.0, retp[1][0], naux);
+        C_DGEMM('T', 'N', naux, naux, navir, 1.0, arQLDp[0], naux, QYarQp[0], naux, 1.0, retp[2][0], naux);
+        C_DGEMM('T', 'N', naux, naux, navir, 1.0, arQp[0], naux, YarQLp[0], naux, 1.0, retp[3][0], naux);
+        C_DGEMM('T', 'N', naux, naux, navir, 1.0, YarQLp[0], naux, QXarQp[0], naux, 1.0, retp[4][0], naux);
+
+    }
+    
+    ret.push_back(Lar);
+    ret.push_back(LDar);
+
+    return ret;
```

Comment:
The more I look at this `std::map<str, SharedMatrix>` feels like a better return type for simplicity and clarity.

Reply:
I think I'm getting this... so this would return as a dict mapping names to matrices to python side. I think I'm still not quite sure how to handle the dict to convert these matrices to numpy arrays though

Reply:
On the python side they will be SharedMatrices. You can use the `np.array(matrix)` or `matrix.np` accessory to get a NumPy view.